cocos/renderer/gfx-gles2/GLES2Context.cpp

```diff
@@ -142,18 +142,106 @@ bool GLES2Context::initialize(const ContextInfo &info) {
         _colorFmt = Format::RGBA8;
         _depthStencilFmt = Format::D24S8;
 
-        const EGLint attribs[] = {
+        bool msaaEnabled = _device->hasFeature(Feature::MSAA);
```

Comment:
This is an incorrect use of the public interface. `hasFeature` serves as a query interface for the upper layer to lookup the feature availabilities. Here the dependency is reversed. I think you need to add a bool property to `DeviceInfo` struct, specifying whether MSAA is desired, and use that here as the request flag. Then after the back buffer initialization processes, return the actual sample count as the evidence for the device feature flag, which the upper layer can now query into.

Reply:
any docs? this made half work in vain